### PR TITLE
8316735: Print LockStack in hs_err files

### PR DESCRIPTION
**Pull request by TheRealMDoerr**
Example output:
```
Lock stack of current Java thread (top to bottom):
LockStack[1]: nsk.share.jdi.EventHandler 
{0x00000000bcc28198} - klass: 'nsk/share/jdi/EventHandler'
 - ---- fields (total size 5 words):
 - private volatile 'wasInterrupted' 'Z' @12  false (0x00)
 - private 'debuggee' 'Lnsk/share/jdi/Debugee;' @16  a 'nsk/share/jdi/LocalLaunchedDebugee'{0x00000000bcc08c18} (0xbcc08c18)
 - private 'log' 'Lnsk/share/Log;' @20  a 'nsk/share/Log'{0x00000000bcc08cb0} (0xbcc08cb0)
 - private 'vm' 'Lcom/sun/jdi/VirtualMachine;' @24  a 'com/sun/tools/jdi/VirtualMachineImpl'{0x00000000bccb3d60} (0xbccb3d60)
 - private 'requestManager' 'Lcom/sun/jdi/request/EventRequestManager;' @28  a 'com/sun/tools/jdi/EventRequestManagerImpl'{0x00000000bccb56f8} (0xbccb56f8)
 - private 'listenThread' 'Ljava/lang/Thread;' @32  a 'java/lang/Thread'{0x00000000bcc280e8} (0xbcc280e8)
LockStack[0]: java.util.Collections$SynchronizedRandomAccessList 
{0x00000000bcb163e8} - klass: 'java/util/Collections$SynchronizedRandomAccessList'
 - ---- fields (total size 3 words):
 - final 'c' 'Ljava/util/Collection;' @12  a 'java/util/Vector'{0x00000000bcb16400} (0xbcb16400)
 - final 'mutex' 'Ljava/lang/Object;' @16  a 'java/util/Collections$SynchronizedRandomAccessList'{0x00000000bcb163e8} (0xbcb163e8)
 - final 'list' 'Ljava/util/List;' @20  a 'java/util/Vector'{0x00000000bcb16400} (0xbcb16400)
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316735](https://bugs.openjdk.org/browse/JDK-8316735): Print LockStack in hs_err files (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15884/head:pull/15884` \
`$ git checkout pull/15884`

Update a local copy of the PR: \
`$ git checkout pull/15884` \
`$ git pull https://git.openjdk.org/jdk.git pull/15884/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15884`

View PR using the GUI difftool: \
`$ git pr show -t 15884`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15884.diff">https://git.openjdk.org/jdk/pull/15884.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15884#issuecomment-1731142950)

### CODE REVIEW

File: src/hotspot/share/runtime/lockStack.hpp [hunk 1]
```diff
@@ -30,8 +30,9 @@
 #include "utilities/globalDefinitions.hpp"
 #include "utilities/sizes.hpp"
 
-class Thread;
+class JavaThread;
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/lockStack.hpp:33]: Seems an unrelated change.

File: src/hotspot/share/utilities/vmError.cpp [hunk 2]
```diff
@@ -1167,6 +1167,10 @@ void VMError::report(outputStream* st, bool _verbose) {
       st->cr();
     }
 
+  STEP_IF("printing fast locked objects", _verbose && _thread != nullptr && _thread->is_Java_thread() && LockingMode == LM_LIGHTWEIGHT);
+    st->print_cr("Objects fast locked by this thread (top to bottom):");
+    JavaThread::cast(_thread)->lock_stack().print_on(st);
+
```
Comment (dholmes-ora) [inline src/hotspot/share/utilities/vmError.cpp:1173]: I would have expected this to be printed along with the other current thread info.

"fast locked" is not terminology we are using for this any more IIUC. I would suggest just saying this is the lock stack for the thread.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Yes it is a good idea to have this. Potentially we may want it for thread dumps too (separate RFE fine).

Thanks

File: src/hotspot/share/runtime/lockStack.hpp (hunk 1 above)
Comment (TheRealMDoerr) [reply src/hotspot/share/runtime/lockStack.hpp:33]: This is a minor cleanup of the prototypes. lockStack.hpp uses `JavaThread*`, not `Thread*`. Do you prefer not to touch it in this PR?

File: src/hotspot/share/utilities/vmError.cpp (hunk 2 above)
Comment (TheRealMDoerr) [reply src/hotspot/share/utilities/vmError.cpp:1173]: I had put it at the end of the "T H R E A D" section. It's moved a bit up, now. Changed wording and added newline.

**Comment by TheRealMDoerr** [comment]
> Yes it is a good idea to have this. Potentially we may want it for thread dumps too (separate RFE fine).
> 
> Thanks

I think that the fast locked objects can already be found by thread dumps by inspecting the frames. But, I haven't checked. Should we file a new issue for investigating?

File: src/hotspot/share/runtime/lockStack.hpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/runtime/lockStack.hpp:33]: If it uses `JavaThread` but only had `Thread` then it must be including it some other way, so the prototype seems unnecessary. But okay to leave it.

**Review by dholmes-ora — APPROVED** [review]
Updates looks good - thanks.

**Comment by dholmes-ora** [comment]
> I think that the fast locked objects can already be found by thread dumps by inspecting the frames. But, I haven't checked. Should we file a new issue for investigating?

I'm not sure how we could present the information in the context of the thread stack dump, even though it might be useful to validate it against the locking information found in the frames themselves. It is probably too disruptive to just jam it into the existing output. We can revisit this later if we find we are missing the information. Thanks.

**Review by MBaesken — APPROVED** [review]
LGTM

**Comment by TheRealMDoerr** [comment]
Thanks for the reviews!
/integrate
